rs-matter/src/transport/network/btp/context.rs

```diff
@@ -0,0 +1,378 @@
+/*
+ *
+ *    Copyright (c) 2020-2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+use core::cell::RefCell;
+
+use embassy_sync::blocking_mutex::{raw::RawMutex, Mutex};
+use log::{error, info, trace, warn};
+
+use crate::error::{Error, ErrorCode};
+use crate::transport::network::BtAddr;
+use crate::utils::notification::Notification;
+
+use super::{session::Session, GattPeripheralEvent};
+
+/// The maximum number of BTP sessions that can be active at any given time.
+/// This is an `rs-matter` specific limit, and is not a requirement of the Matter BTP spec, and which in future should be configurable.
+///
+/// The `GattPeripheral` implementation is expected to enforce this limit as well,
+/// i.e. it should not allow more than `MAX_BTP_SESSIONS` active subscriptions to characteristic `C2`.
+pub const MAX_BTP_SESSIONS: usize = 2;
+
+/// Represents an error that occurred while trying to lock a session for sending.
+#[derive(Debug)]
+pub(crate) struct LockError(());
+
+/// An internal utility for representing a session which is locked for sending.
+///
+/// This type is used to ensure that at any moment in time, a session either is not sending anything,
+/// or is sending the BTP PDUs of a single BTP SDU, which is a requirement of the Matter BTP spec.
+///
+/// The send lock is removed once this object is dropped.
+pub(crate) struct SessionSendLock<'a, M>
+where
+    M: RawMutex,
+{
+    context: &'a BtpContext<M>,
+    address: BtAddr,
+}
+
+impl<'a, M> SessionSendLock<'a, M>
+where
+    M: RawMutex,
+{
+    /// Try to find a session that matches the given condition and lock it for sending.
+    ///
+    /// If there is no session matching the given condition, the methos will return `LockError`.
+    /// If the first session matching the given condition is already locked for sending, the method will return `None`.
```

Comment:
Fair enough, I agree the return type is misleading. I've therefore changed `LockError` to represent an enum with two members as you suggest:
- `NoMatch` - there is no such session
- `AlreadyLocked` - the session is currently locked

W.r.t. the `lock` sibling of `try_lock` that we are discussing here... I've left its signature as-is, but I renamed it to `lock_any` as it does a best effort to find _any_ session which is simultaneously not locked for sending, and in the state of `condition`. I.e. `condition` - for that method - is not unique at all. I.e. it could be "any session that is about to expire and it so happens that we currently don't send stuff over it".